lib/Sema/SemaExpr.cpp

```diff
@@ -8058,6 +8058,112 @@ Sema::CheckSingleAssignmentConstraints(QualType LHSType, ExprResult &CallerRHS,
   return result;
 }
 
+/// If a declaration has a Checked C bounds-safe interface
+/// attached to it, return the checked pointer type for the interface.
+/// Otherwise return a null QualType.
+///
+/// This falls into two cases:
+/// 1. The interface is a type annotation: return the type in the
+///    annotation.
+/// 2. The interface is a bounds expression.  This implies the checked
+/// type should be an _Array_ptr type.  Construct an _Array_ptr version
+/// of the unchecked ponter type for the declaration and return it.
+QualType Sema::GetInteropType(const ValueDecl *Decl) {
+  const DeclaratorDecl *TargetDecl = nullptr;
+  if (const FieldDecl *Field = dyn_cast<FieldDecl>(Decl))
+    TargetDecl = Field;
+  else if (const VarDecl *Var = dyn_cast<VarDecl>(Decl))
+    TargetDecl = Var;
+
+  if (TargetDecl) {
+    if (const BoundsExpr *Bounds = TargetDecl->getBoundsExpr()) {
+      switch (Bounds->getKind()) {
+        case BoundsExpr::Kind::InteropTypeAnnotation: {
+          const InteropTypeBoundsAnnotation *Annot =
+            dyn_cast<InteropTypeBoundsAnnotation>(Bounds);
+          assert(Annot && "unexpected dyn_cast failure");
+          if (Annot != nullptr)
+            return Annot->getType();
```

Comment:
Did you mean to fall-through here?  I know you don't expect `Annot` to be `nullptr`, but if it were, should we bail?